packages/atomic/src/components/common/atomic-modal/atomic-modal.ts

```diff
@@ -0,0 +1,441 @@
+      }
+      this.isOpen = false;
+    }
+  };
+
+  private async handleModalClose() {
+    const executionId = this.createExecutionTracker();
+
+    this.removeModalOpenedClasses();
+    this.setFocusTrapActive(false);
+
+    await this.waitForAnimationEnded();
+    this.onAnimationEnded();
+    if (!this.isExecutionValid(executionId)) {
+      return;
+    }
+
+    this.shouldRender = false;
+  }
+
+  private async handleModalOpen() {
+    const executionId = this.createExecutionTracker();
+
+    if (!this.shouldRender) {
+      this.shouldRender = true;
+      await this.updateComplete;
+    }
+
+    this.addModalOpenedClasses();
+
+    await this.waitForAnimationEnded();
+    if (!this.isExecutionValid(executionId)) {
+      return;
+    }
+
+    this.setFocusTrapActive(true);
+  }
+
+  private async handleToggleOpen(isOpen: boolean) {
+    if (isOpen) {
+      await this.handleModalOpen();
+    } else {
+      await this.handleModalClose();
+    }
+  }
+
+  private isExecutionValid(executionId: number): boolean {
+    return executionId === this.currentWatchToggleOpenExecution;
+  }
+
+  private onWindowTouchMove = (e: Event) => {
+    this.isOpen && e.preventDefault();
+  };
+
+  private removeModalOpenedClasses() {
+    const modalOpenedClass = 'atomic-modal-opened';
+    document.body.classList.remove(modalOpenedClass);
+    this.bindings?.interfaceElement?.classList?.remove(modalOpenedClass);
+  }
+
+  private renderContent() {
+    return html`
+      <article
+        part="container"
+        class="bg-background text-on-background flex flex-col justify-between ${this.isOpen ? 'animate-open' : 'animate-close'}"
+      
+        ${ref(this.animatableContainer)}
+      >
+        <header part="header-wrapper" class="flex flex-col items-center">
+          <div
+            part="header"
+            class="flex justify-between w-full max-w-lg text-xl"
+            id=${this.headerId}
+          >
+            <slot name="header"></slot>
+          </div>
+        </header>
+        <hr part="header-ruler" class="border-t border-neutral"/>
+        <div
+          part="body-wrapper"
+          class="flex w-full grow flex-col items-center overflow-auto [scrollbar-gutter:stable_both-edges]"
+        >
+          <div
+            part="body"
+            class="w-full max-w-lg"
+            @touchmove=${(e: Event) => this.isOpen && e.stopPropagation()}
+          >
+            <slot name="body"></slot>
+          </div>
+        </div>
+        <footer
+          part="footer-wrapper"
+          class="z-10 flex flex-col items-center w-full border-t border-neutral bg-background"
+        >
+          <div part="footer" class="w-full max-w-lg">
+            <slot name="footer"></slot>
+          </div>
+        </footer>
+      </article>
+    `;
+  }
+
+  private setFocusTrapActive(active: boolean) {
+    if (this.focusTrap.value) {
+      this.focusTrap.value.active = active;
+    }
+  }
+
+  private updateHostClasses() {
```

Comment:
I feel like this is an anti-pattern.
Why not use the `:host()` selectors instead of manipulating the DOM.
e.g.
```css
/* Instead of .open class */
:host([is-open]) [part='backdrop'] {
  pointer-events: auto;
}

:host([is-open][dialog]) [part='backdrop'] {
  background-color: rgba(40, 40, 40, 0.8);
}

/* Instead of .dialog class */
:host([dialog]) [part='backdrop'] {
  display: grid;
  padding: 1.5rem;
  /* ... rest of dialog styles */
}

/* Instead of .fullscreen class */
:host([fullscreen]) [part='container'] {
  position: absolute;
  inset: 0;
}
```

Reply:
I like this idea!

The atomic-modal element doesn't have a dialog attribute, though, so I'll use:

```css
:host(:not([fullscreen]))
```
instead.